File: Python3/프로그래머스/2/42839. 소수 찾기/소수 찾기.py
import math
prime_num = set()

def permutation(numbers, visited, N, select):
    for i in range(N):
        if not visited[i]:
            select.append(numbers[i])
            prime_num.add(int(''.join(select)))
            visited[i] = True
            permutation(numbers, visited, N, select)
            visited[i] = False
            select.pop()

def is_prime(number):
    if number <= 1:
        return False
    for i in range(2, int(math.sqrt(number)) + 1):
        if number % i == 0:
            return False
    return True
            
def solution(numbers):
    answer = 0
    numbers = list(numbers)
    permutation(numbers, [False] * len(numbers), len(numbers), [])
    
    return len([i for i in prime_num if is_prime(i)])

File: C++/프로그래머스/2/42839. 소수 찾기/소수 찾기.cpp
#include <string>
#include <vector>
#include <unordered_set>
#include <iostream>
#include <cmath>

using namespace std;

unordered_set<int> pm_num;
vector<string> numbers;
vector<bool> visited;
int N;

void permutation(string selected) {
    for(int i = 0; i < N; i++) {
        if(!visited[i]) {
            visited[i] = true;
            pm_num.insert(stoi(selected + numbers[i]));
            permutation(selected + numbers[i]);
            visited[i] = false;
        }
    }
}

bool is_prime(int number) {
    if(number <= 1) return false;
    for(int i = 2; i <= sqrt(number); i++) {
        if(number % i == 0) {
            return false;
        }
    }
    return true;
}

int solution(string numbers) {
    int answer = 0;
    ::N = numbers.size();
    ::visited = vector<bool>(N);
    for(int i = 0; i < numbers.size(); i++) {
        ::numbers.push_back(numbers.substr(i, 1));
    }
    permutation("");
    
    for(auto& num : ::pm_num) {
        if(is_prime(num)) answer++;
    }
    
    return answer;
}